auth_totp/models/res_users.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 LasLabs Inc.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from datetime import datetime, timedelta
+import random
+import string
+from openerp import api, fields, models
+from openerp.exceptions import AccessDenied, ValidationError
+from ..exceptions import MfaTokenInvalidError, MfaTokenExpiredError
+
+
+class ResUsers(models.Model):
+    _inherit = 'res.users'
+
+    mfa_enabled = fields.Boolean(string='MFA Enabled?')
+    authenticator_ids = fields.One2many(
+        comodel_name='res.users.authenticator',
+        inverse_name='user_id',
+        string='Authentication Apps/Devices',
+        help='To delete an authentication app, remove it from this list. To'
+        ' add a new authentication app, please use the button to the right.',
+    )
+    mfa_login_token = fields.Char()
+    mfa_login_token_exp = fields.Datetime()
+    trusted_device_ids = fields.One2many(
+        comodel_name='res.users.device',
+        inverse_name='user_id',
+    )
+
+    @api.multi
+    @api.constrains('mfa_enabled', 'authenticator_ids')
+    def _check_enabled_with_authenticator(self):
+        for record in self:
+            if record.mfa_enabled and not record.authenticator_ids:
+                raise ValidationError(
+                    'You have MFA enabled but do not have any authentication'
+                    ' apps/devices set up. To keep from being locked out,'
+                    ' please add one before you activate this feature.'
+                )
+
+    @api.model
+    def check_credentials(self, password):
+        try:
+            super(ResUsers, self).check_credentials(password)
+        except AccessDenied:
+            user = self.sudo().search([
+                ('id', '=', self.env.uid),
+                ('mfa_login_token', '=', password),
+            ])
+            user._user_from_mfa_login_token_validate()
```

Comment:
There's no explicit return in the original method: https://github.com/odoo/odoo/blob/9.0/openerp/addons/base/res/res_users.py#L442